cmake/CompilerWarnings.cmake

```diff
@@ -63,7 +63,6 @@ function(set_file_warnings)
         -Woverloaded-virtual # warn if you overload (not override) a virtual function
         -Wconversion # warn on type conversions that may lose data
         -Wsign-conversion # warn on sign conversions
-        -Wdouble-promotion # warn if float is implicit promoted to double
```

Comment:
Why remove this? Sure, there's technically no loss of precision, but this is not so uncontroversial. Java allows implicit widening conversions, while Kotlin and Rust do not.

Reply:
This is from the extended commit message from the commit that made this change:

> Promoting from `float` to `double` is a lossless operation. We do
not need to introduce a `static_cast` whenever this happens. Lossy
conversions back to `float` are easily caught by `-Wconversion`. I
don't think this warning adds enough value to justify the increased
verbosity and decreased readabilty of all the additional casts it
requires we write.

One particular problem is that we keep this warning, we must add many dozen `static_cast<double>(someFLoat)` calls to the tests which drastically impacts readability while providing no extra correctness.

Reply:
Having an `Approx` helper that works on float would indeed solve this cleanly.

Reply:
Right, maybe we have to write our own wrappers around `doctest::Approx(double)` to work around this. I wasn't excited to remove `-Wdouble-promotion` and would prefer to keep it if it didn't mean destroying test legibility. However I'm also not excited to write wrappers around the test framework since I already think our testing infrastructure is growing too complicated.

Reply:
I reintroduced this warning. Luckily it wasn't as hard as I thought to keep it.